File: src/java/Quicksort.java
/*
    The program expects a python list in a textfile as input.
    Avg. runtime of Quickssort: O(n*log(n))
*/

import java.util.Arrays;
import java.util.Scanner;

public class Quicksort {

    private static int[] numbers;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();
        scanner.close();

        String[] inputString = input.replaceAll("\\[|\\]|\\s", "").split(",");
        numbers = new int[inputString.length];
        for (int i=0; i<inputString.length; i++) {
            numbers[i] = Integer.parseInt(inputString[i]);
        }

        quicksort(0, numbers.length-1);
        System.out.println(Arrays.toString(numbers));
    }

    public static int divide(int l, int r) {
        int p = numbers[(l + r) / 2];
        while (true) {
            while (numbers[l] < p)
                l++;
            while (numbers[r] > p)
                r--;
            if (l < r) {
                int tmp = numbers[l];
                numbers[l] = numbers[r];
                numbers[r] = tmp;
            } else
                return r;
        }
    }

    public static void quicksort(int l, int r) {
        if (l < r) {
            int d = divide(l, r);
            quicksort(l, d);
            quicksort(d+1, r);
        }
    }
}

File: src/python/Quicksort.py
"""
    The program expects a python list in a textfile as input.
    Avg. runtime of Quickssort: O(n*log(n))
"""

import ast

def quicksort(l, r):
        if (l < r):
            d = divide(l, r)
            quicksort(l, d)
            quicksort(d+1, r)

def divide(l, r):
    p = numbers[int((l + r) / 2)]
    while (1):
        while (numbers[l] < p):
            l = l+1
        while (numbers[r] > p):
            r = r-1
        if (l < r):
            tmp = numbers[l]
            numbers[l] = numbers[r]
            numbers[r] = tmp
        else:
            return r

numbers = input()
numbers = ast.literal_eval(numbers)

quicksort(0, len(numbers)-1)
print(numbers)